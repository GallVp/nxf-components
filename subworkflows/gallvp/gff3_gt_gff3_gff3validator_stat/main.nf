include { GT_GFF3                               } from '../../../modules/gallvp/gt/gff3/main'
include { SAMTOOLS_FAIDX                        } from '../../../modules/gallvp/samtools/faidx/main'
include { GT_GFF3VALIDATOR                      } from '../../../modules/gallvp/gt/gff3validator/main'
include { GT_STAT                               } from '../../../modules/gallvp/gt/stat/main'

workflow GFF3_GT_GFF3_GFF3VALIDATOR_STAT {

    take:
    ch_gff3                                     // channel: [ val(meta), gff3 ]
    ch_fasta                                    // channel: [ val(meta), fasta ]

    main:

    ch_versions                                 = Channel.empty()

    // MODULE: GT_GFF3
    GT_GFF3 ( ch_gff3 )

    ch_versions                                 = ch_versions.mix(GT_GFF3.out.versions.first())

    // MODULE: GT_GFF3VALIDATOR
    GT_GFF3VALIDATOR ( GT_GFF3.out.gt_gff3 )

    ch_versions                                 = ch_versions.mix(GT_GFF3VALIDATOR.out.versions.first())

    // MODULE: SAMTOOLS_FAIDX
    SAMTOOLS_FAIDX(
        ch_fasta,
        [ [], [] ]
    )

    ch_fai                                      = SAMTOOLS_FAIDX.out.fai
    ch_versions                                 = ch_versions.mix(SAMTOOLS_FAIDX.out.versions.first())

    // FUNCTION: checkGff3FastaCorrespondence
    ch_gff3_fai                                 = GT_GFF3VALIDATOR.out.success_log
                                                | join (
                                                    GT_GFF3.out.gt_gff3
                                                )
                                                | map { meta, log, gff3 -> [ meta, gff3 ] }
                                                | join (
                                                    ch_fai
                                                )

    ch_correspondence_status                    = ch_gff3_fai
                                                | map { meta, gff3, fai ->
                                                    checkGff3FastaCorrespondence ( meta, gff3, fai )
                                                }

    ch_correspondence_success                   = ch_correspondence_status
                                                | map { meta, success, error ->
                                                    if ( success ) {
                                                        [ meta, success ]
                                                    }
                                                }

    ch_correspondence_error                     = ch_correspondence_status
                                                | map { meta, success, error ->
                                                    if ( error ) {
                                                        [ "${meta.id}.error.log", error.join('\n') ]
                                                    }
                                                }
                                                | collectFile
                                                | map { error ->
                                                    [ [ id: error.baseName.replace('.error', '') ], error ]
                                                }

    ch_valid_gff3                               = ch_correspondence_success
                                                | join (
                                                    ch_gff3_fai
                                                    | map { meta, gff3, fai -> [ meta, gff3 ] }
                                                )
                                                | map { meta, log, gff3 -> [ meta, gff3 ] }

    ch_log_for_invalid_gff3                     = GT_GFF3.out.error_log
                                                | mix (
                                                    GT_GFF3VALIDATOR.out.error_log
                                                )
                                                | mix (
                                                    ch_correspondence_error
                                                )

    // MODULE: GT_STAT
    GT_STAT ( ch_valid_gff3 )

    ch_gff3_stats                               = GT_STAT.out.stats
    ch_versions                                 = ch_versions.mix(GT_STAT.out.versions.first())

    emit:
    valid_gff3              = ch_valid_gff3             // channel: [ val(meta), gff3 ]
    gff3_stats              = ch_gff3_stats             // channel: [ val(meta), yml ]
    log_for_invalid_gff3    = ch_log_for_invalid_gff3   // channel: [ val(meta), log ]
    versions                = ch_versions               // channel: [ versions.yml ]
}

def checkGff3FastaCorrespondence(meta, gff3File, faiFile) {

    // STEP 1
    // Check that gff3 has no identifiers that are not in fasta
    def gff3Lines           = gff3File.readLines().findAll { ! it.startsWith('#') }
    def gff3Identifiers     = gff3Lines.collect { it.split('\t')[0] }.unique().sort()
    def fastaIdentifiers    = faiFile.readLines().collect { it.split('\t')[0] }.unique().sort()
    def missingIdentifiers  = gff3Identifiers.findAll { ! fastaIdentifiers.contains(it) }

    if (missingIdentifiers) {
        return [
            meta,
            [], // success log
            [
                "Failed to validate gff3 file: ${gff3File.name}",
                "GFF3 file contains identifiers not present in FASTA:",
                "${missingIdentifiers.join('\n')}"
            ] // error log
        ]
    }

    // STEP 2
    // Check that there are no coordinates in gff3 that exceed the sequence length in the parent fasta entry
    def sequenceLengths = [:]
    faiFile.readLines().each { line ->
        def parts = line.split('\t')
        sequenceLengths[parts[0]] = parts[1]
    }

    for ( line in gff3Lines ) {
        def parts = line.split('\t')
        def name = parts[0]
        def start = parts[3].toInteger()
        def end = parts[4].toInteger()
        def seqLength = sequenceLengths[name].toInteger()

        if (start > seqLength || end > seqLength) {
            return [
                meta,
                [], // success log
                [
                    "Failed to validate gff3: ${gff3File.name}",
                    "Coordinates exceed sequence length in GFF3 file:",
                    "Sequence: $name",
                    "Sequence length: $seqLength",
                    "Start: $start",
                    "End: $end"
                ] // error log
            ]
        }
    }

    return [
        meta,
        [
            "All tests passed..."
        ], // success log
        [] // error log
    ]
}
