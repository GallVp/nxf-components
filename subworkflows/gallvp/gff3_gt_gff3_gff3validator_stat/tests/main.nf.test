nextflow_workflow {

    name "Test Workflow GFF3_GT_GFF3_GFF3VALIDATOR_STAT"
    script "../main.nf"
    workflow "GFF3_GT_GFF3_GFF3VALIDATOR_STAT"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/gff3_gt_gff3_gff3validator_stat"
    tag "gff3_gt_gff3_gff3validator_stat"
    tag "gt"
    tag "gt/gff3"
    tag "gt/gff3validator"
    tag "gt/stat"
    tag "samtools"
    tag "samtools/faidx"

    test("sarscov2-genome_gff3-all_pass") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gff3', checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("homo_sapiens-genome_bed-gt_gff3_fail") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.bed', checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2-genome_gff3-homo_sapiens-genome_fasta-correspondence_fail") {

        when {
            workflow {
                """
                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gff3', checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.valid_gff3,
                    workflow.out.versions).match()
                },
                { assert path(workflow.out.log_for_invalid_gff3[0][1]).text.contains('GFF3 file contains identifiers not present in FASTA') }
            )
        }
    }

    test("sarscov2-genome_gff3-test_fasta-correspondence_fail-out_of_bounds") {

        when {
            workflow {
                """
                def test_fasta = new File('genome.fasta')
                test_fasta.text = '>MT192765.1\\nAGCTAGCT'

                input[0] = Channel.of([ [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.gff3', checkIfExists: true)
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    test_fasta.toPath()
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.valid_gff3,
                    workflow.out.versions).match()
                },
                { assert path(workflow.out.log_for_invalid_gff3[0][1]).text.contains('Start coordinates exceed sequence length in the GFF3 file') }
            )
        }
    }

    test("sarscov2 - fasta - circular_region - pass") {

        when {
            workflow {
                """
                def circular_gff = new File('circular_gff.gff')
                circular_gff.text = [
                    '##gff-version 3',
                    'MT192765.1	Genbank	region	1	29829	.	+	.	circular=true',
                    'MT192765.1	Genbank	gene	29551	39667	.	+	.	ID=gene1',
                    'MT192765.1	Genbank	CDS	29551	39667	.	+	0	Parent=gene1'
                ].join('\\n')

                input[0] = Channel.of([ [ id:'test' ], // meta map
                    circular_gff.toPath()
                ])
                input[1] = Channel.of([ [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}