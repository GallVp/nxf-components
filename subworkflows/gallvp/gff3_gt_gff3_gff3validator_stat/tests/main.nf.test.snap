{
    "homo_sapiens-genome_bed-gt_gff3_fail": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,c096494c3cd02864eb54434c294ba382"
                    ]
                ],
                "3": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106"
                ],
                "gff3_stats": [
                    
                ],
                "log_for_invalid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.error.log:md5,c096494c3cd02864eb54434c294ba382"
                    ]
                ],
                "valid_gff3": [
                    
                ],
                "versions": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T16:21:54.482267"
    },
    "sarscov2-genome_gff3-test_fasta-correspondence_fail-out_of_bounds": {
        "content": [
            [
                
            ],
            [
                "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T16:22:06.684959"
    },
    "sarscov2 - fasta - circular_region - pass": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,b3bb01b18b8eeac28922ab55c5c6c939"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.yml:md5,545b8e290cfa8a93fd0ff01ad9daee08"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,80555fe6e28e9564cb534f5478842286",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                    "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
                ],
                "gff3_stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test.yml:md5,545b8e290cfa8a93fd0ff01ad9daee08"
                    ]
                ],
                "log_for_invalid_gff3": [
                    
                ],
                "valid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,b3bb01b18b8eeac28922ab55c5c6c939"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,80555fe6e28e9564cb534f5478842286",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                    "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-19T13:53:32.901064"
    },
    "sarscov2-genome_gff3-homo_sapiens-genome_fasta-correspondence_fail": {
        "content": [
            [
                
            ],
            [
                "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T16:22:00.684573"
    },
    "sarscov2-genome_gff3-all_pass": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,2ae900237ace415557b8735fac088b85"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.yml:md5,762d733791108d9723989c3bbc69fa2b"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,80555fe6e28e9564cb534f5478842286",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                    "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
                ],
                "gff3_stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test.yml:md5,762d733791108d9723989c3bbc69fa2b"
                    ]
                ],
                "log_for_invalid_gff3": [
                    
                ],
                "valid_gff3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.gt.gff3:md5,2ae900237ace415557b8735fac088b85"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0cb9519e626e5128d8495cf29b7d59ff",
                    "versions.yml:md5,80555fe6e28e9564cb534f5478842286",
                    "versions.yml:md5,8a418ac34d045b0cdac812eb2dc9c106",
                    "versions.yml:md5,c89b081a13c68acc5326e43ca9104344"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T16:21:49.138904"
    }
}